pkgs/build-support/replace-dependencies.nix

```diff
@@ -0,0 +1,140 @@
+{ lib, runCommandLocal, replaceDirectDependencies }:
+
+# Replace some dependencies in the requisites tree of drv, propagating the change all the way up the tree, even within other replacements, without a full rebuild.
+# This can be useful, for example, to patch a security hole in libc and still use your system safely without rebuilding the world.
+# This should be a short term solution, as soon as a rebuild can be done the properly rebuilt derivation should be used.
+# Each old dependency and the corresponding new dependency MUST have the same-length name, and ideally should have close-to-identical directory layout.
+#
+# Example: safeFirefox = replaceDependencies {
+#   drv = firefox;
+#   replacements = [
+#     {
+#       oldDependency = glibc;
+#       newDependency = glibc.overrideAttrs (oldAttrs: {
+#         patches = oldAttrs.patches ++ [ ./fix-glibc-hole.patch ];
+#       });
+#     }
+#     {
+#       oldDependency = libwebp;
+#       newDependency = libwebp.overrideAttrs (oldAttrs: {
+#         patches = oldAttrs.patches ++ [ ./fix-libwebp-hole.patch ];
+#       });
+#     }
+#   ];
+# };
+# This will first rebuild glibc and libwebp with your security patches.
+# Then it copies over firefox (and all of its dependencies) without rebuilding further.
+# In particular, the glibc dependency of libwebp will be replaced by the patched version as well.
+#
+# In rare cases, it is possible for the replacement process to cause breakage (for example due to checksum mismatch).
+# The cutoffPackages argument can be used to exempt the problematic packages from the replacement process.
+{ drv, replacements, cutoffPackages ? [ ], verbose ? true }:
```

Comment:
Question: should the `cutoffPackages` default to `map (x: x.newDependency) replacements`, as in the case of the backward-compatible `replaceDependency`?

Reply:
Certainly not, because it completely nullifies one of the two big advantages of `replaceDependencies` (namely, replacing dependencies in other replacements as well); in fact it would even make it worse because the old workaround of choosing the order carefully does not work any more for the NixOS option.